flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java

```diff
@@ -845,21 +845,19 @@ public ExecutorService getAsyncOperationsThreadPool() {
 
 	private void notifyCheckpointComplete(long checkpointId) {
 		try {
-			boolean success = actionExecutor.call(() -> {
```

Comment:
I guess we can also get ride of `actionExecutor` from `#performCheckpoint` method which is actually executed by mailbox thread.

Reply:
Maybe I misunderstood the scope for `actionExecutor`? I thought it is only used for executing operations for non-task/mailbox thread for unification. And I checked all the usages for `performCheckpoint`, and it seems all from task thread execution. If it is called from barrier handler via `triggerCheckpointOnBarrier`, it should be still from task thread? If so we will not need to delegate it to `actionExecutor`.

Reply:
`actionExecutor` was added during checkpoint lock deprecation to unify `SourceStreamTask` and `StreamTask`. 
Basically, under `StreamTask` it's a no-op operation going through `StreamTaskActionExecutor.IMMEDIATE`, while `SourceStreamTask` is using `StreamTaskActionExecutor#SynchronizedStreamTaskActionExecutor`, which acquires checkpoint lock for all operations.
I agree that the naming and the tight integration into the `MailboxExecutor` makes it hard to understand at first. Luckily, we can get rid of it, once FLIP-27 is fully implemented and we can get rid of legacy source thread (so in about 3 years? ;)).